RFC-0023-Unified-Memory-for-Pytorch.md

```diff
@@ -0,0 +1,127 @@
+
+   - With a single allocator everything is more easily tracked, and this design more closely aligns with the managed memory API usage model. 
+
+### Empty Tensors
+Changes will be needed for the `at::empty_cpu()` and `at::empty_cuda()` functions. 
+
+ - In `at::empty_cpu()`, there is already a switch to determine which cpu allocator to use (pinned vs cpu). When UVM is enabled, this switch will always return the CachingManagedAllocator. There are likely a few other places (other empties) that get an allocator as well.
+
+ - Per a find from the FBGEMM team, there is an issue with the context being moved from the device where the memory was allocated, and where it currently resides. We need to keep the original context, so we don’t create new ones unnecessarily on other devices. To account for this, at `at::empty` time, we will create two Storage structs: 
+   - RealStorage -- This will be the original context of the device where the CUDAMallocManaged occurred. 
+   - Indirect -- This will be the new Storage that contains the new context of where the memory is currently being used, with reference back to the release of the RealStorage being moved with it.  (This is like the shared pointer’s design.) 
+
+### Tensor Copies
+In the UVM model, the managed memory APIs can be relied upon to ensure data and tensors are available and accessible and explicit calls to `cudaMemCopy()`/`hipMemCopy()` are not required for moving tensors to a GPU device. Users, however, often expect to deal with copies. 
+
+### UVM Copies vs Traditional
+Currently when PyTorch performs an update to a tensor’s attribute(s), a copy of the tensor is returned to the caller that includes the requested change. In a UVM case, this is potentially suboptimal. For example, when the user only wants to ensure the tensor is transferred to a GPU and issues a tensor.to(device=”cuda”) with a change only to the device attribute, this should not require an additional Block allocation and potential `CudaMallocManaged()` call to accomplish. 
+
+On the other hand, if a conversion of a tensor under UVM changes any other attribute (dtype, layout, etc) the tensor will need to make decisions based on the attributes of the current and requested tensor. This will require a traditional call to the `_to_copy` function to return the expected object to the user. 
+
+We propose to treat tensor.to(device) as a special case when UVM is enabled and `enabled_uvm_move_on_copy` has been set. Additionally, the behavior of tensor.uvm_to() will match the behavior of tensor.to() when UVM is enabled and  `enabled_uvm_move_on_copy` is True.
+
+![Existing and proposed allocator usage](./RFC-0023-assets/copy-diagram.jpg)
```

Comment:
I am not sure what it means here for the cpu Tensor not to exist anymore? the copy op is not modifying the source inplace so I am not sure how this behavior would fit within pytorch.

Does this mean that there is no way to have `b = a.to("cuda")` (with a on cpu initially) and have:
- a will keep behaving as a cpu Tensor
- b will behave as a gpu Tensor
- a and b share memory (in the sense that inplace ops on either of them is reflected on the other).

Effectively, this means that this `.to()` will only be a cross-device view op (meaning that the input and output share memory). The parallel to this in core is the `torch.view_as_real()` function for example that is a view op for which the input is complex and the output is float dtype.

Reply:
In line with the statement I made up above, we are going to add the use of the `copy=true` parameter on the `.to()` function to allow an explicit copy to be made even when the tensor is managed by uvm.  

For the `torch.view_as_real()` we are treating a `.to()` as an explicit change of DeviceType for the tensor.  Even though the memory address may be the same, `.to()` will trigger a prefetch to tell the driver to start moving the memory to the target device.  Necessary synchronizations and other actions will be taken as needed to ensure correct coherency.